README.md
# ROS Kinesis Service Common Library


## Overview
This repository contains the kinesis_manager package, a library for interacting with Amazon Kinesis Video Streams. It is used by the `kinesis_video_streamer` package.

### License
The source code is released under an [Apache 2.0].

**Author**: AWS RoboMaker<br/>
**Affiliation**: [Amazon Web Services (AWS)]<br/>
**Maintainer**: AWS RoboMaker, ros-contributions@amazon.com

### Supported ROS Distributions
- Kinetic
- Melodic
- Dashing

### Build status
* Travis CI:
    * "master" branch [![Build Status](https://travis-ci.org/aws-robotics/kinesisvideo-common.svg?branch=master)](https://travis-ci.org/aws-robotics/kinesisvideo-common/branches)
    * "release-latest" branch [![Build Status](https://travis-ci.org/aws-robotics/kinesisvideo-common.svg?branch=release-latest)](https://travis-ci.org/aws-robotics/kinesisvideo-common/branches)
* ROS build farm:
    * ROS Kinetic @ u16.04 Xenial [![Build Status](http://build.ros.org/job/Kbin_uX64__kinesis_manager__ubuntu_xenial_amd64__binary/badge/icon)](http://build.ros.org/job/Kbin_uX64__kinesis_manager__ubuntu_xenial_amd64__binary)
    * ROS Melodic @ u18.04 Bionic [![Build Status](http://build.ros.org/job/Mbin_uB64__kinesis_manager__ubuntu_bionic_amd64__binary/badge/icon)](http://build.ros.org/job/Mbin_uB64__kinesis_manager__ubuntu_bionic_amd64__binary/)
   * ROS Dashing @ u18.04 Bionic [![Build Status](http://build.ros2.org/job/Dbin_uB64__kinesis_manager__ubuntu_bionic_amd64__binary/badge/icon)](http://build.ros2.org/job/Dbin_uB64__kinesis_manager__ubuntu_bionic_amd64__binary)

## Installation

### Binaries
On Ubuntu you can install the latest version of this package using the following command

        sudo apt-get update
        sudo apt-get install -y ros-$ROS_DISTRO-kinesis-manager

### Building from Source

To build from source you'll need to create a new workspace, clone and checkout the latest release branch of this repository, install all the dependencies, and compile. If you need the latest development features you can clone from the `master` branch instead of the latest release branch. While we guarantee the release branches are stable, __the `master` should be considered to have an unstable build__ due to ongoing development. 

- Install build tool: please refer to `colcon` [installation guide](https://colcon.readthedocs.io/en/released/user/installation.html)

- Create a ROS workspace and a source directory

        mkdir -p ~/ros-workspace/src

- Clone the package into the source directory . 

        cd ~/ros-workspace/src
        git clone https://github.com/aws-robotics/kinesisvideo-common.git -b release-latest

- Install dependencies

        cd ~/ros-workspace 
        sudo apt-get update && rosdep update
        rosdep install --from-paths src --ignore-src -r -y
        
_Note: If building the master branch instead of a release branch you may need to also checkout and build the master branches of the packages this package depends on._

- Build the packages

        cd ~/ros-workspace && colcon build

- Configure ROS library Path

        source ~/ros-workspace/install/local_setup.bash


[Amazon Web Services (AWS)]: https://aws.amazon.com/
[Apache 2.0]: https://aws.amazon.com/apache-2-0/


kinesis_manager/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package kinesis_manager
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------
* Bump patch version to 2.0.3 for release (`#40 <https://github.com/aws-robotics/kinesisvideo-common/issues/40>`_)
* Patch the kvs sdk to not 'using namespace std' in its headers. (`#39 <https://github.com/aws-robotics/kinesisvideo-common/issues/39>`_)
  * Patch the kvs sdk to not 'using namespace std' in its headers.
  Resolves potential namespace collision with other libraries.
  Backpatch of https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/pull/541 to v1.7.8 that we are using.
* Link libraries only if test targets exist (`#36 <https://github.com/aws-robotics/kinesisvideo-common/issues/36>`_)
  * Link libraries only if test targets exist
  * Increment version to 2.0.2
* Add test cases for empty region / callback providers in InitializeVideoProducer.
* Add logging for missing region
* Merge pull request `#32 <https://github.com/aws-robotics/kinesisvideo-common/issues/32>`_ from aws-robotics/bump-version-2-0-1
  Bump version
* Bump version
  md5 sum was changed. Bumping version in preparation for release.
* Update md5 checksum for KVS SDK (`#27 <https://github.com/aws-robotics/kinesisvideo-common/issues/27>`_)
* Use standard CMake macros for adding gtest/gmock tests (`#24 <https://github.com/aws-robotics/kinesisvideo-common/issues/24>`_)
  * use the macro in aws_common to find test dependencies for ROS1 or ROS2
  * update travis.yml to be compatible with specifying multiple package names
  * update travis.yml test matrix
  * update PACKAGE_NAMES
* update kvs producer sdk to v1.7.8
* Improve unit tests coverage
  * Add KinesisVideoProducerInterface and
  KinesisVideoStreamInterface in order to use
  mocks for Kinesis Video Streams
  * Resulting in the following overall coverage rate:
  ```
  lines......: 88.8% (906 of 1020 lines)
  functions..: 86.0% (208 of 242 functions)
  branches...: 38.0% (1346 of 3544 branches)
  ```
* Update to use non-legacy ParameterReader API (`#12 <https://github.com/aws-robotics/kinesisvideo-common/issues/12>`_)
* Update to use new ParameterReader API (`#11 <https://github.com/aws-robotics/kinesisvideo-common/issues/11>`_)
  * clean up CMakeFiles
  * Adjusting usage of the ParameterReader due to the updated API taking in a ParameterPath.
  * refactor based on new ParameterPath object design
  * increment major version number in package.xml
* Merge pull request `#5 <https://github.com/aws-robotics/kinesisvideo-common/issues/5>`_ from mm318/master
  Executable permission for script files
* fix permission issues when this repo is zipped
* Contributors: AAlon, Avishay Alon, Emerson Knapp, Juan Rodriguez Hortala, M. M, Miaofei, Ross Desmond, Tim Robinson


kinesis_manager/test/default_callbacks_test.cpp
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include <kinesis_manager/default_callbacks.h>
#include <gtest/gtest.h>

using namespace Aws::Kinesis;

TEST(ClientCallbackProviderSuite, defaultClientCallbackProviderTest)
{
  Aws::Kinesis::DefaultClientCallbackProvider test_subject;
  UINT64 custom_handle;
  UINT64 remaining_bytes;

  EXPECT_EQ(test_subject.storageOverflowPressure(custom_handle, remaining_bytes), 
    test_subject.getStorageOverflowPressureCallback()(custom_handle, remaining_bytes));
  EXPECT_EQ(STATUS_SUCCESS, 
    test_subject.storageOverflowPressure(custom_handle, remaining_bytes));
}

TEST(ClientCallbackProviderSuite, defaultStreamCallbackProviderTest)
{
  DefaultStreamCallbackProvider test_subject;
  UINT64 custom_data;
  STREAM_HANDLE stream_handle;
  UINT64 last_buffering_ack;
  UINT64 errored_timecode;
  STATUS status_code;
  UINT64 dropped_frame_timecode;

  EXPECT_EQ(STATUS_SUCCESS,
    test_subject.getStreamConnectionStaleCallback()(custom_data, stream_handle, last_buffering_ack));
  EXPECT_EQ(STATUS_SUCCESS,
    test_subject.getStreamErrorReportCallback()(custom_data, stream_handle, errored_timecode, status_code));
  EXPECT_EQ(STATUS_SUCCESS,
    test_subject.getDroppedFrameReportCallback()(custom_data, stream_handle, dropped_frame_timecode));
}

int main(int argc, char ** argv)
{
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

kinesis_manager/src/default_callbacks.cpp
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
#include <aws/core/utils/logging/LogMacros.h>
#include <kinesis_manager/default_callbacks.h>

using namespace std;
using namespace com::amazonaws::kinesis::video;


namespace Aws {
namespace Kinesis {

/*******************************
 * Client callbacks
 *******************************/
UINT64 DefaultClientCallbackProvider::getCallbackCustomData()
{
  return reinterpret_cast<UINT64>(this);
}
StorageOverflowPressureFunc DefaultClientCallbackProvider::getStorageOverflowPressureCallback()
{
  return storageOverflowPressure;
}
STATUS DefaultClientCallbackProvider::storageOverflowPressure(UINT64 custom_handle,
                                                              UINT64 remaining_bytes)
{
  UNUSED_PARAM(custom_handle);
  return STATUS_SUCCESS;
}

/*******************************
 * Credentials setup
 *******************************/
unique_ptr<CredentialProvider> CreateDefaultCredentialProvider()
{
  char const *access_key, *secret_key, *session_token;
  if (nullptr == (access_key = getenv(AWS_ACCESS_KEY_ENV_VAR))) {
    return unique_ptr<CredentialProvider>{};
  }
  if (nullptr == (secret_key = getenv(AWS_SECRET_KEY_ENV_VAR))) {
    return unique_ptr<CredentialProvider>{};
  }
  if (nullptr == (session_token = getenv(AWS_SESSION_TOKEN_ENV_VAR))) {
    session_token = "";
  }

  auto now_time = std::chrono::duration_cast<std::chrono::seconds>(
    std::chrono::system_clock::now().time_since_epoch());
  unique_ptr<Credentials> credentials_ = make_unique<Credentials>(
    string(access_key), string(secret_key), string(session_token),
    now_time + std::chrono::seconds(AWS_DEFAULT_CREDENTIAL_ROTATION_PERIOD_IN_SECONDS));
  return make_unique<StaticCredentialProvider>(*credentials_.get());
}

/*******************************
 * Stream callbacks
 *******************************/
UINT64 DefaultStreamCallbackProvider::getCallbackCustomData()
{
  return reinterpret_cast<UINT64>(this);
}

StreamConnectionStaleFunc DefaultStreamCallbackProvider::getStreamConnectionStaleCallback()
{
  return streamConnectionStaleHandler;
};

StreamErrorReportFunc DefaultStreamCallbackProvider::getStreamErrorReportCallback()
{
  return streamErrorReportHandler;
};

DroppedFrameReportFunc DefaultStreamCallbackProvider::getDroppedFrameReportCallback()
{
  return droppedFrameReportHandler;
};

STATUS DefaultStreamCallbackProvider::streamConnectionStaleHandler(UINT64 custom_data,
                                                                   STREAM_HANDLE stream_handle,
                                                                   UINT64 last_buffering_ack)
{
  AWS_LOGSTREAM_WARN(__func__, "Reporting stream stale. Last ACK received " << last_buffering_ack);
  return STATUS_SUCCESS;
}

STATUS
DefaultStreamCallbackProvider::streamErrorReportHandler(UINT64 custom_data,
                                                        STREAM_HANDLE stream_handle,
                                                        UINT64 errored_timecode, STATUS status_code)
{
  AWS_LOGSTREAM_ERROR(__func__, "Reporting stream error. Errored timecode: "
                                  << errored_timecode << " Status: " << status_code);
  return STATUS_SUCCESS;
}

STATUS
DefaultStreamCallbackProvider::droppedFrameReportHandler(UINT64 custom_data,
                                                         STREAM_HANDLE stream_handle,
                                                         UINT64 dropped_frame_timecode)
{
  AWS_LOGSTREAM_WARN(__func__,
                     "Reporting dropped frame. Frame timecode " << dropped_frame_timecode);
  return STATUS_SUCCESS;
}

}  // namespace Kinesis
}  // namespace Aws

kinesis_manager/include/kinesis_manager/common.h
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
#pragma once
#include <sstream>

#define KINESIS_MANAGER_STATUS_SUCCEEDED(status) (!(KINESIS_MANAGER_STATUS_FAILED(status)))
#define KINESIS_MANAGER_STATUS_FAILED(status) \
  ((status)&Aws::Kinesis::KINESIS_MANAGER_STATUS_ERROR_BASE)

#ifndef INVALID_STREAM_ID
#define INVALID_STREAM_ID (-1)
#endif

namespace Aws {
namespace Kinesis {

typedef enum kinesis_manager_status_e {
  KINESIS_MANAGER_STATUS_SUCCESS = 0,
  KINESIS_MANAGER_STATUS_ERROR_BASE = 0x1000,
  KINESIS_MANAGER_STATUS_BASE64DECODE_FAILED,                         /* 0x1001 */
  KINESIS_MANAGER_STATUS_MALLOC_FAILED,                               /* 0x1002 */
  KINESIS_MANAGER_STATUS_INVALID_INPUT,                               /* 0x1003 */
  KINESIS_MANAGER_STATUS_CREATESTREAMSYNC_FAILED,                     /* 0x1004 */
  KINESIS_MANAGER_STATUS_PUTFRAME_STREAM_NOT_FOUND,                   /* 0x1005 */
  KINESIS_MANAGER_STATUS_PUTFRAME_FAILED,                             /* 0x1006 */
  KINESIS_MANAGER_STATUS_PUTMETADATA_STREAM_NOT_FOUND,                /* 0x1007 */
  KINESIS_MANAGER_STATUS_PUTMETADATA_FAILED,                          /* 0x1008 */
  KINESIS_MANAGER_STATUS_VIDEO_PRODUCER_NOT_INITIALIZED,              /* 0x1009 */
  KINESIS_MANAGER_STATUS_VIDEO_PRODUCER_ALREADY_INITIALIZED,          /* 0x100A */
  KINESIS_MANAGER_STATUS_DEFAULT_CREDENTIAL_PROVIDER_CREATION_FAILED, /* 0x100B */
  KINESIS_MANAGER_STATUS_EMPTY_STREAM_NAME,                           /* 0x100C */
  KINESIS_MANAGER_STATUS_STREAM_ALREADY_EXISTS,                       /* 0x100D */
  KINESIS_MANAGER_STATUS_SUBSCRIPTION_INSTALLATION_FAILED,            /* 0x100E */
  KINESIS_MANAGER_STATUS_SUBSCRIPTION_INSTALLER_NOT_FOUND,            /* 0x100F */
  KINESIS_MANAGER_STATUS_PROCESS_CODEC_DATA_STREAM_CONFIG_NOT_FOUND,  /* 0x1010 */
  KINESIS_MANAGER_STATUS_GET_STREAM_DEFINITION_FAILED,                /* 0x1011 */
  KINESIS_MANAGER_STATUS_NO_STREAMS_SPECIFIED,                        /* 0x1012 */
  KINESIS_MANAGER_STATUS_LIST_SHARDS_FAILED,                          /* 0x1013 */
  KINESIS_MANAGER_STATUS_LIST_SHARDS_EMPTY,                           /* 0x1014 */
  KINESIS_MANAGER_STATUS_GET_SHARD_ITERATOR_FAILED,                   /* 0x1015 */
  KINESIS_MANAGER_STATUS_GET_RECORDS_FAILED,                          /* 0x1016 */
  KINESIS_MANAGER_STATUS_GET_RECORDS_THROTTLED,                       /* 0x1017 */
} KinesisManagerStatus;

typedef int KinesisStreamInputType;
constexpr uint32_t kDefaultMessageQueueSize = 1000;
constexpr uint16_t kDefaultRecordsLimitForRekognitionResults = 50;

/**
 * The following structure represents the format in which parameters for the streamer node are
 * expected. kinesis_video/ stream_count: int stream<idx>/ subscription_topic: string
 *   subscription_queue_size: optional int, defaults to 100.
 *   topic_type: int in range of KinesisStreamInputType
 *   < StreamDefinition parameters as defined in
 * https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-data.html#how-data-header-streamdefinition
 * >
 *  ... more streams ...
 */
const struct
{
  const char * prefix;
  const char * stream_count;
  const char * stream_namespace;
  const char * topic_name;
  const char * message_queue_size;
  const char * topic_type;
  const char * log4cplus_config;
  const char * stream_name;
  const char * rekognition_topic_name;
  const char * rekognition_data_stream;
} kStreamParameters{
  "kinesis_video/",
  "stream_count",
  "stream",
  "subscription_topic",
  "subscription_queue_size", /* Overrides kDefaultMessageQueueSize */
  "topic_type",              /* Topic type value as defined in KinesisStreamInputType */
  "log4cplus_config",        /* Path to a log4cplus configuration file for use by the Kinesis Video
                                Producer SDK */
  "stream_name",
  "rekognition_topic_name", /* AWS Rekognition analysis results will be published to this topic. */
  "rekognition_data_stream" /* AWS Rekognition analysis results will be read from this stream. Must
                               be provided if topic_type is Rekognition-enabled. */
};

}  // namespace Kinesis
}  // namespace Aws

kinesis_manager/include/kinesis_manager/default_callbacks.h
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
/**
 * @brief This module provides default callbacks for use with the Kinesis Producer Library.
 *  You can implement your own custom callbacks or use the default ones declared here.
 * @see https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/producer-reference-callbacks.html
 */
#pragma once
#include <aws/core/auth/AWSCredentialsProviderChain.h>
#include <kinesis-video-producer/Auth.h>
#include <kinesis-video-producer/ClientCallbackProvider.h>
#include <kinesis-video-producer/DefaultCallbackProvider.h>
#include <kinesis-video-producer/DefaultDeviceInfoProvider.h>

#define AWS_ACCESS_KEY_ENV_VAR "AWS_ACCESS_KEY_ID"
#define AWS_SECRET_KEY_ENV_VAR "AWS_SECRET_ACCESS_KEY"
#define AWS_SESSION_TOKEN_ENV_VAR "AWS_SESSION_TOKEN"
#define AWS_DEFAULT_CREDENTIAL_ROTATION_PERIOD_IN_SECONDS (2400)


namespace Aws {
namespace Kinesis {

/**
 * Creates a static CredentialProvider holding a single credential pair with expiration set to
 * AWS_DEFAULT_CREDENTIAL_ROTATION_PERIOD_IN_SECONDS.. The credentials will be loaded from
 * environment variables matching the AWS SDK.
 * @return CredentialProvider
 */
std::unique_ptr<com::amazonaws::kinesis::video::CredentialProvider> CreateDefaultCredentialProvider();
/**
 * Credentials provider which uses the AWS SDK's default credential provider chain.
 * @note You need to have called Aws::InitAPI before using this provider.
 */
class ProducerSdkAWSCredentialsProvider : public com::amazonaws::kinesis::video::CredentialProvider
{
public:
  ProducerSdkAWSCredentialsProvider(std::shared_ptr<Auth::DefaultAWSCredentialsProviderChain>
                                      default_aws_credentials_provider = nullptr)
  {
    if (default_aws_credentials_provider) {
      default_aws_credentials_provider_ = default_aws_credentials_provider;
    } else {
      default_aws_credentials_provider_ =
        Aws::MakeShared<Auth::DefaultAWSCredentialsProviderChain>(__func__);
    }
  }

private:
  std::shared_ptr<Auth::DefaultAWSCredentialsProviderChain> default_aws_credentials_provider_;

  void updateCredentials(com::amazonaws::kinesis::video::Credentials & producer_sdk_credentials) override
  {
    Auth::AWSCredentials aws_sdk_credentials =
      default_aws_credentials_provider_->GetAWSCredentials();
    producer_sdk_credentials.setAccessKey(aws_sdk_credentials.GetAWSAccessKeyId().c_str());
    producer_sdk_credentials.setSecretKey(aws_sdk_credentials.GetAWSSecretKey().c_str());
    producer_sdk_credentials.setSessionToken(aws_sdk_credentials.GetSessionToken().c_str());
    auto now = std::chrono::duration_cast<std::chrono::seconds>(
      std::chrono::system_clock::now().time_since_epoch());
    auto refresh_interval = std::chrono::duration_cast<std::chrono::seconds>(
      std::chrono::milliseconds(Auth::REFRESH_THRESHOLD));
    producer_sdk_credentials.setExpiration(now + refresh_interval);
  }
};

class DefaultClientCallbackProvider : public com::amazonaws::kinesis::video::ClientCallbackProvider
{
public:
  UINT64 getCallbackCustomData() override;
  StorageOverflowPressureFunc getStorageOverflowPressureCallback() override;
  static STATUS storageOverflowPressure(UINT64 custom_handle, UINT64 remaining_bytes);
};

class DefaultStreamCallbackProvider : public com::amazonaws::kinesis::video::StreamCallbackProvider
{
public:
  UINT64 getCallbackCustomData() override;
  StreamConnectionStaleFunc getStreamConnectionStaleCallback() override;
  StreamErrorReportFunc getStreamErrorReportCallback() override;
  DroppedFrameReportFunc getDroppedFrameReportCallback() override;

private:
  static STATUS streamConnectionStaleHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
                                             UINT64 last_buffering_ack);
  static STATUS streamErrorReportHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
                                         UINT64 errored_timecode, STATUS status_code);
  static STATUS droppedFrameReportHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
                                          UINT64 dropped_frame_timecode);
};

}  // namespace Kinesis
}  // namespace Aws

kinesis_manager/include/kinesis_manager/stream_definition_provider.h
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
#pragma once
#include <aws_common/sdk_utils/client_configuration_provider.h>
#include <kinesis-video-producer/KinesisVideoProducer.h>
#include <kinesis_manager/common.h>


namespace Aws {
namespace Kinesis {
/**
 * Responsible for creating stream definition objects by using a parameter reader to get the
 * relevant data.
 * @see https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-data.html
 */
class StreamDefinitionProvider
{
public:
  ~StreamDefinitionProvider() = default;
  /**
   * Reads a base64-encoded codecPrivateData using the given Parameter Reader.
   * @param prefix a ParameterPath object representing the path of the parameter's prefix.
   * @param reader implementing ParameterReaderInterface
   * @param out_codec_private_data - newly allocated buffer containing the codec private data.
   *  @note The pointer remains unchanged in case of an error or if no codec data was provided.
   *  @note The caller of GetCodecPrivateData is responsible for freeing up this buffer.
   * @param out_codec_private_data_size - the actual size of the allocated buffer will be stored
   * here.
   * @return kinesis_manager_status_t with KINESIS_MANAGER_SUCCESS on success or if no codec data
   * was provided
   */
  virtual KinesisManagerStatus
  GetCodecPrivateData(const Aws::Client::ParameterPath & prefix,
                      const Aws::Client::ParameterReaderInterface & reader,
                      PBYTE * out_codec_private_data,
                      uint32_t * out_codec_private_data_size) const;

  /**
   * Creates a StreamDefinition by using a given Parameter Reader. Uses sensible defaults in case
   * the parameter is not configured.
   * @see
   * https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/master/kinesis-video-producer/src/StreamDefinition.h
   *  and https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-data.html
   * @param prefix a ParameterPath object representing the path of the parameter's prefix.
   * @param reader implementing ParameterReaderInterface
   * @param codec_private_data the picture and sequence parameter sets of the stream (optional).
   * @param codec_private_data_size size in bytes of codec_private_data.
   * @return unique_ptr to StreamDefinition on success or nullptr on failure.
   */
  virtual unique_ptr<com::amazonaws::kinesis::video::StreamDefinition>
  GetStreamDefinition(const Aws::Client::ParameterPath & prefix,
                      const Aws::Client::ParameterReaderInterface & reader,
                      const PBYTE codec_private_data,
                      uint32_t codec_private_data_size) const;
};

}  // namespace Kinesis
}  // namespace Aws


kinesis_manager/include/kinesis_manager/stream_subscription_installer.h
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
#pragma once
#include <aws_common/sdk_utils/logging/aws_log_system.h>
#include <kinesis_manager/common.h>

#include <map>

namespace Aws {
namespace Kinesis {

struct StreamSubscriptionDescriptor
{
  KinesisStreamInputType input_type;
  std::string topic_name;
  std::string stream_name;
  uint32_t message_queue_size;
  std::string rekognition_topic_name;
  std::string rekognition_data_stream;
};

typedef std::function<bool(const StreamSubscriptionDescriptor & descriptor)> SubscriberSetupFn;

/**
 * This class abstracts away the installation of media sources (i.e. subscriptions) for Kinesis
 * streams.
 */
class StreamSubscriptionInstaller
{
public:
  StreamSubscriptionInstaller() = default;
  virtual ~StreamSubscriptionInstaller() = default;
  /**
   * Installs a subscription that will provide and handle the stream's input. This function is a
   * shim layer that maps an input type to a pre-set installer function. Child classes should
   * implement the individual installers.
   * @param input_type
   * @param topic_name
   * @param stream_name
   * @param message_queue_size
   * @return KinesisManagerStatus
   */
  virtual KinesisManagerStatus Install(const StreamSubscriptionDescriptor & descriptor) const
  {
    if (descriptor.topic_name.empty() || descriptor.stream_name.empty()) {
      return KINESIS_MANAGER_STATUS_INVALID_INPUT;
    }
    if (0 == installers_.count(descriptor.input_type)) {
      return KINESIS_MANAGER_STATUS_SUBSCRIPTION_INSTALLER_NOT_FOUND;
    }

    bool result = installers_.at(descriptor.input_type)(descriptor);
    return result ? KINESIS_MANAGER_STATUS_SUCCESS
                  : KINESIS_MANAGER_STATUS_SUBSCRIPTION_INSTALLATION_FAILED;
  }
  /**
   * Uninstalls all subscriptions for the given topic name.
   * @param topic_name
   */
  virtual void Uninstall(const std::string & topic_name) = 0;

protected:
  /**
   * Different stream input types will require different subscriber setup functions.
   */
  std::map<KinesisStreamInputType, SubscriberSetupFn> installers_;
};

}  // namespace Kinesis
}  // namespace Aws


kinesis_manager/src/stream_definition_provider.cpp
/*
 * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
#include <aws_common/sdk_utils/client_configuration_provider.h>
#include <com/amazonaws/kinesis/video/utils/Include.h>
#include <kinesis_manager/stream_definition_provider.h>

using namespace com::amazonaws::kinesis::video;
using namespace Aws::Client;


#define STREAM_DEFINITION_MAX_CODEC_PRIVATE_DATA_SIZE (1024)


namespace Aws {
namespace Kinesis {

KinesisManagerStatus StreamDefinitionProvider::GetCodecPrivateData(
  const ParameterPath & prefix, const ParameterReaderInterface & reader,
  PBYTE * out_codec_private_data, uint32_t * out_codec_private_data_size) const
{
  if (nullptr == out_codec_private_data ||
      nullptr == out_codec_private_data_size) {
    return KINESIS_MANAGER_STATUS_INVALID_INPUT;
  }
  std::string b64_encoded_codec_private_data;
  reader.ReadParam(prefix + "codecPrivateData", b64_encoded_codec_private_data);
  if (!b64_encoded_codec_private_data.empty()) {
    uint8_t temp_codec_data[STREAM_DEFINITION_MAX_CODEC_PRIVATE_DATA_SIZE] = {0};
    uint32_t decoded_buffer_size = sizeof(temp_codec_data);
    if (STATUS_SUCCESS != base64Decode(const_cast<char *>(b64_encoded_codec_private_data.c_str()),
                                       temp_codec_data, &decoded_buffer_size)) {
      return KINESIS_MANAGER_STATUS_BASE64DECODE_FAILED;
    }
    PBYTE codec_private_data = (PBYTE)malloc(decoded_buffer_size);
    if (nullptr == codec_private_data) {
      return KINESIS_MANAGER_STATUS_MALLOC_FAILED;
    }
    memset(codec_private_data, 0, decoded_buffer_size);
    memcpy(codec_private_data, temp_codec_data, decoded_buffer_size);
    *out_codec_private_data = codec_private_data;
    *out_codec_private_data_size = decoded_buffer_size;
  }
  return KINESIS_MANAGER_STATUS_SUCCESS;
}

unique_ptr<StreamDefinition> StreamDefinitionProvider::GetStreamDefinition(
    const ParameterPath & prefix, const ParameterReaderInterface & reader,
    const PBYTE codec_private_data, uint32_t codec_private_data_size) const
{
  if (nullptr == codec_private_data && 0 != codec_private_data_size) {
    return unique_ptr<StreamDefinition>{};
  }

  std::string stream_name = "default";
  reader.ReadParam(prefix + "stream_name", stream_name);

  map<string, string> tags;
  reader.ReadParam(prefix + "tags", tags);

  int retention_period = 2;
  reader.ReadParam(prefix + "retention_period", retention_period);

  std::string kms_key_id;
  reader.ReadParam(prefix + "kms_key_id", kms_key_id);

  int streaming_type_id = 0;
  reader.ReadParam(prefix + "streaming_type", streaming_type_id);
  STREAMING_TYPE streaming_type = static_cast<STREAMING_TYPE>(streaming_type_id);

  std::string content_type = "video/h264";
  reader.ReadParam(prefix + "content_type", content_type);


  int max_latency = 0;
  reader.ReadParam(prefix + "max_latency", max_latency);

  int fragment_duration = 2;
  reader.ReadParam(prefix + "fragment_duration", fragment_duration);

  int timecode_scale = 1;
  reader.ReadParam(prefix + "timecode_scale", timecode_scale);

  bool key_frame_fragmentation = true;
  reader.ReadParam(prefix + "key_frame_fragmentation", key_frame_fragmentation);

  bool frame_timecodes = true;
  reader.ReadParam(prefix + "frame_timecodes", frame_timecodes);

  bool absolute_fragment_time = true;
  reader.ReadParam(prefix + "absolute_fragment_time", absolute_fragment_time);

  bool fragment_acks = true;
  reader.ReadParam(prefix + "fragment_acks", fragment_acks);

  bool restart_on_error = true;
  reader.ReadParam(prefix + "restart_on_error", restart_on_error);

  bool recalculate_metrics = true;
  reader.ReadParam(prefix + "recalculate_metrics", recalculate_metrics);

  int nal_adaptation_flag_id = NAL_ADAPTATION_ANNEXB_NALS | NAL_ADAPTATION_ANNEXB_CPD_NALS;
  reader.ReadParam(prefix + "nal_adaptation_flags", nal_adaptation_flag_id);
  NAL_ADAPTATION_FLAGS nal_adaptation_flags =
    static_cast<NAL_ADAPTATION_FLAGS>(nal_adaptation_flag_id);

  int frame_rate = 24;
  reader.ReadParam(prefix + "frame_rate", frame_rate);

  int avg_bandwidth_bps = 4 * 1024 * 1024;
  reader.ReadParam(prefix + "avg_bandwidth_bps", avg_bandwidth_bps);

  int buffer_duration = 120;
  reader.ReadParam(prefix + "buffer_duration", buffer_duration);

  int replay_duration = 40;
  reader.ReadParam(prefix + "replay_duration", replay_duration);

  int connection_staleness = 30;
  reader.ReadParam(prefix + "connection_staleness", connection_staleness);

  std::string codec_id = "V_MPEG4/ISO/AVC";
  reader.ReadParam(prefix + "codec_id", codec_id);

  std::string track_name = "kinesis_video";
  reader.ReadParam(prefix + "track_name", track_name);


  auto stream_definition = make_unique<StreamDefinition>(
    stream_name, hours(retention_period), &tags, kms_key_id, streaming_type, content_type,
    milliseconds(max_latency), seconds(fragment_duration), milliseconds(timecode_scale),
    key_frame_fragmentation, frame_timecodes, absolute_fragment_time, fragment_acks,
    restart_on_error, recalculate_metrics, nal_adaptation_flags, frame_rate, avg_bandwidth_bps,
    seconds(buffer_duration), seconds(replay_duration), seconds(connection_staleness), codec_id,
    track_name, codec_private_data, codec_private_data_size);
  return stream_definition;
}
}  // namespace Kinesis
}  // namespace Aws
